src/jetstream_push_subscription.rs

```diff
@@ -0,0 +1,544 @@
+        }
+    }
+
+    /// Returns a blocking message iterator.
+    /// Same as calling `iter()`.
+    ///
+    /// # Example
+    /// ```no_run
+    /// # fn main() -> std::io::Result<()> {
+    /// # let nc = nats::connect("demo.nats.io")?;
+    /// # let sub = nc.subscribe("foo")?;
+    /// for msg in sub.messages() {}
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn messages(&self) -> Iter<'_> {
+        Iter { subscription: self }
+    }
+
+    /// Returns a blocking message iterator.
+    ///
+    /// # Example
+    /// ```no_run
+    /// # fn main() -> std::io::Result<()> {
+    /// # let nc = nats::connect("demo.nats.io")?;
+    /// # let sub = nc.subscribe("foo")?;
+    /// for msg in sub.iter() {}
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn iter(&self) -> Iter<'_> {
+        Iter { subscription: self }
+    }
+
+    /// Returns a non-blocking message iterator.
+    ///
+    /// # Example
+    /// ```
+    /// # fn main() -> std::io::Result<()> {
+    /// # let nc = nats::connect("demo.nats.io")?;
+    /// # let sub = nc.subscribe("foo")?;
+    /// for msg in sub.try_iter() {}
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn try_iter(&self) -> TryIter<'_> {
+        TryIter { subscription: self }
+    }
+
+    /// Returns a blocking message iterator with a time
+    /// deadline for blocking.
+    ///
+    /// # Example
+    /// ```
+    /// # fn main() -> std::io::Result<()> {
+    /// # let nc = nats::connect("demo.nats.io")?;
+    /// # let sub = nc.subscribe("foo")?;
+    /// for msg in sub.timeout_iter(std::time::Duration::from_secs(1)) {}
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn timeout_iter(&self, timeout: Duration) -> TimeoutIter<'_> {
+        TimeoutIter {
+            subscription: self,
+            to: timeout,
+        }
+    }
+
+    /// Process and acknowledge a single message, waiting indefinitely for
+    /// one to arrive.
+    ///
+    /// Does not acknowledge the processed message if the closure returns an `Err`.
+    ///
+    /// Does not return an `Err` if acking the message is unsuccessful. If you require stronger
+    /// processing guarantees, you can manually call the `double_ack` method of the argument
+    /// message.
+    ///
+    /// Example
+    ///
+    /// ```no_run
+    /// # fn main() -> std::io::Result<()> {
+    /// # let nc = nats::connect("demo.nats.io")?;
+    /// # let js = nats::jetstream::new(nc);
+    /// # js.add_stream("foo")?;
+    /// let mut sub = js.subscribe("foo")?;
+    /// # js.publish("foo", "hello")?;
+    /// sub.process(|msg| {
+    ///     println!("Received message {:?}", msg);
+    ///
+    ///     Ok(())
+    /// })?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn process<R, F: Fn(&Message) -> io::Result<R>>(&mut self, f: F) -> io::Result<R> {
+        let next = self.next().unwrap();
+
+        let result = f(&next)?;
+        if self.0.consumer_ack_policy != AckPolicy::None {
+            next.ack().ok();
```

Comment:
I know that it's coming from old implementation of `process` but I'm not conviced on mapping `ack()` to Option and discarding the error telling the user to use `double_ack` (or `ack_sync` in our case) instead. After all, double_ack is different mechanism (using request-response), not just `ack` with error handled.

Reply:
Yeah I just brought this over as it was.

Tangental to this, I wondered if we should just do `with_handler` Ala `Subscription::with_handler` and then also an `with_ack_handler` that always acknowledges.

Reply:
That might work here, since pull is always manual ack IIRC.

Reply:
Want to change `process` calls to error out @Jarema?

Reply:
Changed to error out if ack fails.